<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Running a Workflow in Terra | Immuno Workflow Course – Terra Walkthrough</title>
  <meta name="description" content="Chapter 2 Running a Workflow in Terra | Immuno Workflow Course – Terra Walkthrough" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Running a Workflow in Terra | Immuno Workflow Course – Terra Walkthrough" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Running a Workflow in Terra | Immuno Workflow Course – Terra Walkthrough" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>

<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>




<link rel="stylesheet" href="assets/style_ITN.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://www.itcrtraining.org/"><img src="assets/ITN_logo.png" style="padding-left: 15px; padding-top: 8px;"</a>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> <strong>About</strong></a></li>
<li class="chapter" data-level="2" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html"><i class="fa fa-check"></i><b>2</b> Running a Workflow in Terra</a>
<ul>
<li class="chapter" data-level="2.1" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#prerequisites"><i class="fa fa-check"></i><b>2.1</b> Prerequisites</a></li>
<li class="chapter" data-level="2.2" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#workspace"><i class="fa fa-check"></i><b>2.2</b> Workspace</a></li>
<li class="chapter" data-level="2.3" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#adding-data"><i class="fa fa-check"></i><b>2.3</b> Adding Data</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#uploading-via-terra-ui"><i class="fa fa-check"></i><b>2.3.1</b> Uploading via Terra UI</a></li>
<li class="chapter" data-level="2.3.2" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#uploading-via-command-line"><i class="fa fa-check"></i><b>2.3.2</b> Uploading via Command Line</a></li>
<li class="chapter" data-level="2.3.3" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#a-note-about-sharing-data-across-workspaces."><i class="fa fa-check"></i><b>2.3.3</b> A note about sharing data across workspaces.</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#populating-terra-tables"><i class="fa fa-check"></i><b>2.4</b> Populating Terra Tables</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#adding-the-sequence-table."><i class="fa fa-check"></i><b>2.4.1</b> Adding the “sequence” table.</a></li>
<li class="chapter" data-level="2.4.2" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#adding-the-sample-table."><i class="fa fa-check"></i><b>2.4.2</b> Adding the Sample table.</a></li>
<li class="chapter" data-level="2.4.3" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#adding-the-analysis-table."><i class="fa fa-check"></i><b>2.4.3</b> Adding the Analysis table.</a></li>
<li class="chapter" data-level="2.4.4" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#linking-to-sequences-in-the-analysis-table."><i class="fa fa-check"></i><b>2.4.4</b> Linking to sequences in the analysis table.</a></li>
<li class="chapter" data-level="2.4.5" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#formatting-a-sequence-entry-in-the-sequence-table."><i class="fa fa-check"></i><b>2.4.5</b> Formatting a sequence entry in the sequence table.</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#setting-up-the-immuno.wdl-workflow."><i class="fa fa-check"></i><b>2.5</b> Setting up the <code>immuno.wdl</code> workflow.</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#pulling-the-immuno.wdl-workflow-from-dockstore"><i class="fa fa-check"></i><b>2.5.1</b> Pulling the <code>immuno.wdl</code> workflow from Dockstore</a></li>
<li class="chapter" data-level="2.5.2" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#selecting-the-workflow-data.-1"><i class="fa fa-check"></i><b>2.5.2</b> Selecting the workflow data.</a></li>
<li class="chapter" data-level="2.5.3" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#setting-the-workflow-options.-1"><i class="fa fa-check"></i><b>2.5.3</b> Setting the workflow options.</a></li>
<li class="chapter" data-level="2.5.4" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#setting-the-workflow-inputs.-1"><i class="fa fa-check"></i><b>2.5.4</b> Setting the workflow inputs.</a></li>
<li class="chapter" data-level="2.5.5" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#setting-the-workflow-outputs.-1"><i class="fa fa-check"></i><b>2.5.5</b> Setting the workflow outputs.</a></li>
<li class="chapter" data-level="2.5.6" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#launching-the-workflow"><i class="fa fa-check"></i><b>2.5.6</b> Launching the workflow</a></li>
<li class="chapter" data-level="2.5.7" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#viewing-the-results"><i class="fa fa-check"></i><b>2.5.7</b> Viewing the results</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="running-a-workflow-in-terra.html"><a href="running-a-workflow-in-terra.html#further-information"><i class="fa fa-check"></i><b>2.6</b> Further information</a></li>
</ul></li>
<li class="divider"></li>
<p style="text-align:center;"> <a href="https://github.com/jhudsl/OTTR_Template" target="blank" > This content was published with</a> <a href="https://bookdown.org/" target="blank"> bookdown by: </a> </p>
<p style="text-align:center;"> <a href="https://griffithlab.org/"> The Griffith Lab </a></p>
<p style="text-align:center; font-size: 12px;"> <a href="https://github.com/rstudio4edu/rstudio4edu-book/"> Style adapted from: rstudio4edu-book </a> <a href ="https://creativecommons.org/licenses/by/2.0/"> (CC-BY 2.0) </a></p>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"><strong>Immuno Workflow Course – Terra Walkthrough</strong></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<head>
  <meta name="viewport" content="width=device-width,minimum-scale=1.0,maximum-scale=1.0,initial-scale=1.0">
  <!--script src="https://kit.fontawesome.com/6a26f47516.js"></script-->
  <!--<script src="assets/hideOutput.js"></script>-->
  <link href="assets/style_ITN.css" rel="stylesheet">
</head>
        


<div class="hero-image-container"> 
  <img class= "hero-image" src="assets/itcr_main_image.png">
</div>
<div id="running-a-workflow-in-terra" class="section level1" number="2">
<h1><span class="header-section-number">Chapter 2</span> Running a Workflow in Terra</h1>
<div class="figure">
<img src="resources/images/immuno_on_terra_logo.png" alt="" />
<p class="caption">Immuno on Terra</p>
</div>
<div id="prerequisites" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Prerequisites</h2>
<p>To complete this walkthrough you’ll need to have access to a billing account in GCP. The <a href="https://github.com/wustl-oncology/cloud-workflows/blob/main/docs/getting_started_gcp.md">cloud-workflows repo has more information on setting this up</a>. Then, if you haven’t used Terra before, you’ll have to create a Terra billing project. Terra has <a href="https://support.terra.bio/hc/en-us/articles/360026182251">detailed documentation on setting this up</a>.</p>
</div>
<div id="workspace" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Workspace</h2>
<p>All work in Terra happens in a workspace. Your workspaces are listed on the <a href="https://app.terra.bio/#workspaces">workspaces page</a>. It also has a button to create a workspace–the plus to the right of the word Workspaces.</p>
<div class="figure">
<img src="resources/images/terra_screens/workspaces-heading.png" alt="" />
<p class="caption">Workspaces heading with + button</p>
</div>
<p>A unique and memorable name will make it easier to find later. This walkthrough will use “demonstrating-immuno-workflow-in-terra”. After choosing a name, a billing project must be selected.</p>
<div class="figure">
<img src="resources/images/terra_screens/workspace-create_form.png" alt="" />
<p class="caption">terra workspace form</p>
</div>
</div>
<div id="adding-data" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Adding Data</h2>
<p>The immuno workflow requires BAM or FASTQ files to be available as inputs. This walkthrough will use the HCC1395 data available here:</p>
<ul>
<li><a href="http://genomedata.org/pmbio-workshop/fastqs/all/Exome_Norm.tar">HCC1395 Normal</a></li>
<li><a href="http://genomedata.org/pmbio-workshop/fastqs/all/Exome_Tumor.tar">HCC1395 Tumor</a></li>
<li><a href="http://genomedata.org/pmbio-workshop/fastqs/all/RNAseq_Tumor.tar">HCC1395 Tumor RNA</a></li>
</ul>
<p>These files will need to be extracted and uploaded to Terra. First extract all of these tar files to a <code>sequence_data</code> directory. Then, there are a few ways to upload that directory to Terra:</p>
<div id="uploading-via-terra-ui" class="section level3" number="2.3.1">
<h3><span class="header-section-number">2.3.1</span> Uploading via Terra UI</h3>
<p>On the “Data” tab for the workspace will be a “Files” link under “Other Data”. From that section there is an “Upload” button that can be used to upload the desired data.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-tab_files.png" alt="" />
<p class="caption">terra upload form</p>
</div>
</div>
<div id="uploading-via-command-line" class="section level3" number="2.3.2">
<h3><span class="header-section-number">2.3.2</span> Uploading via Command Line</h3>
<p>On the “Dashboard” tab for the workspace will be a link to the Google Cloud Storage Bucket for the workspace. By copying this bucket URL, the files can be uploaded using the <code>gsutil</code> tool.</p>
<div class="figure">
<img src="resources/images/terra_screens/dashboard-cloud_information.png" alt="" />
<p class="caption">dashboard cloud information</p>
</div>
</div>
<div id="a-note-about-sharing-data-across-workspaces." class="section level3" number="2.3.3">
<h3><span class="header-section-number">2.3.3</span> A note about sharing data across workspaces.</h3>
<p>If the same data will be used by many people, it may be beneficial to upload it to a workspace and configure it once. Thereafter that workspace can be cloned to use the data in another workspace without copying it again thereby saving storage costs. The data tables configured below could also be cloned to the new workspace.</p>
</div>
</div>
<div id="populating-terra-tables" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Populating Terra Tables</h2>
<p>For this example, there will be three data tables: “sequence” (to list the information for each pair of input FASTQs), “sample” (for any metadata about the samples), and “analysis” (to group together the sequences and samples for one workflow run). In the simplest case a workflow can be run from a single table (or by supplying an input JSON directly); the additional tables here are to demonstrate some of the linking features of Terra.</p>
<div id="adding-the-sequence-table." class="section level3" number="2.4.1">
<h3><span class="header-section-number">2.4.1</span> Adding the “sequence” table.</h3>
<p>On the “Data” tab of the workspace is an “Import Data” button. Clicking on it offers a few options–in this case use the import TSV option.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-upload_tsv.png" alt="" />
<p class="caption">import menu</p>
</div>
<p>This offers the option to upload a file or to copy and paste the TSV. To create the “sequence” table, start with the following:</p>
<pre><code>entity:sequence_id  sample_id   fastq1  fastq2  readgroup
1   HCC1395_TUMOR_DNA   gs://BUCKET/sequence_data/Exome_Tumor/Exome_Tumor_R1.fastq.gz   gs://BUCKET/sequence_data/Exome_Tumor/Exome_Tumor_R2.fastq.gz   @RG\tID:1\tPU:C1TD1ACXX\tSM:HCC1395_TUMOR_DNA\tLB:HCC1395_TUMOR_DNA_LIB1\tPL:Illumina\tCN:WUGSC
2   HCC1395_NORMAL_DNA  gs://BUCKET/sequence_data/Exome_Norm/Exome_Norm_R1.fastq.gz gs://BUCKET/sequence_data/Exome_Norm/Exome_Norm_R2.fastq.gz @RG\tID:2\tPU:C1TD1ACXX\tSM:HCC1395_NORMAL_DNA\tLB:HCC1395_NORMAL_DNA_LIB1\tPL:Illumina\tCN:WUGSC
3   HCC1395_TUMOR_RNA   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane1_R1.fastq.gz   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane1_R2.fastq.gz   &quot;ID:3   PU:H3MYFBBXX.4  SM:HCC1395_TUMOR_RNA    LB:HCC1395_TUMOR_RNA_LIB1   PL:Illumina CN:WUGSC&quot;
4   HCC1395_TUMOR_RNA   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane2_R1.fastq.gz   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane2_R2.fastq.gz   &quot;ID:4   PU:H3MYFBBXX.5  SM:HCC1395_TUMOR_RNA    LB:HCC1395_TUMOR_RNA_LIB1   PL:Illumina CN:WUGSC&quot;</code></pre>
<p>The bucket paths need to be updated to point to the uploaded files for this workspace by filling in the <code>BUCKET</code>. Notice the headers of this file… the first column starts with <code>entity:</code> to indicate this column identifies the table. Since it says <code>sequence_id</code> the table will be named “sequence”. For this table, the file import must be used as the text uploader does not handle fields that contain tabs.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-import_file.png" alt="" />
<p class="caption">import file</p>
</div>
</div>
<div id="adding-the-sample-table." class="section level3" number="2.4.2">
<h3><span class="header-section-number">2.4.2</span> Adding the Sample table.</h3>
<p>This time the text import can be used by copying this block:</p>
<pre><code>entity:sample_id    individual_id   type
HCC1395_TUMOR_DNA   HCC1395 tumor_dna
HCC1395_NORMAL_DNA  HCC1395 normal_dna
HCC1395_TUMOR_RNA   HCC1395 tumor_rna</code></pre>
<p>This will create a “sample” table.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-import_text.png" alt="" />
<p class="caption">import text</p>
</div>
</div>
<div id="adding-the-analysis-table." class="section level3" number="2.4.3">
<h3><span class="header-section-number">2.4.3</span> Adding the Analysis table.</h3>
<p>As before, open the TSV import and copy this block:</p>
<pre><code>entity:analysis_id  individual_id   tumor_dna_sample    normal_dna_sample   tumor_rna_sample
HCC1395 HCC1395 {&quot;entityType&quot;:&quot;sample&quot;,&quot;entityName&quot;:&quot;HCC1395_NORMAL_DNA&quot;}   {&quot;entityType&quot;:&quot;sample&quot;,&quot;entityName&quot;:&quot;HCC1395_TUMOR_DNA&quot;}    {&quot;entityType&quot;:&quot;sample&quot;,&quot;entityName&quot;:&quot;HCC1395_TUMOR_RNA&quot;}</code></pre>
<p>Notice this table has some entities in the columns. This will automatically link those entries to the “sample” table. This allows access to properties of the “sample” table via this table. If there were another table for individuals, the “sample” table could’ve been created with links to that table for further nested indirect access.</p>
</div>
<div id="linking-to-sequences-in-the-analysis-table." class="section level3" number="2.4.4">
<h3><span class="header-section-number">2.4.4</span> Linking to sequences in the analysis table.</h3>
<p>Now that the basic tables are in place, there are a few more steps to do before using the workflow. First is to add columns to the analysis table to point to which sequence applies. From the “analysis” table, there is an edit button with an option to “Add Column”. Use it three times, one for each sample.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-edit_analysis_table.png" alt="" />
<p class="caption">edit analysis table</p>
</div>
<p>Each time, use a type of “reference” to the “sequence” table and select the option that the values are a list.</p>
<p>The three columns to add and their values:</p>
<table>
<thead>
<tr class="header">
<th>column</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>tumor_dna_sequences</td>
<td>1</td>
</tr>
<tr class="even">
<td>normal_dna_sequences</td>
<td>2</td>
</tr>
<tr class="odd">
<td>tumor_rna_sequences</td>
<td>3,4</td>
</tr>
</tbody>
</table>
<p>The option to use a list is selected every time because the workflow expects a list of data, even if some lists only contain one item.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-add_analysis_column.png" alt="" />
<p class="caption">add column form</p>
</div>
<p>As a note, although this walkthrough used the Terra UI to demonstrate adding a column to the workflow, it would’ve been possible to create these entities in the TSV directly and skip this step so long as the column is formatted correctly. An example of how to format a list in the TSV:</p>
<pre><code>[{&quot;entityType&quot;:&quot;sequence&quot;,&quot;entityName&quot;:&quot;1&quot;},{&quot;entityType&quot;:&quot;sequence&quot;,&quot;entityName&quot;:&quot;2&quot;}]</code></pre>
</div>
<div id="formatting-a-sequence-entry-in-the-sequence-table." class="section level3" number="2.4.5">
<h3><span class="header-section-number">2.4.5</span> Formatting a sequence entry in the sequence table.</h3>
<p>The Terra interface does not support adding a JSON object directly as a single column unless that object is the output of a workflow. Therefore a separate zero-step WDL workflow is needed in order to format the sequence objects in the way the <code>immuno.wdl</code> expects. This is a great opportunity to see a simple case of setting up a workflow before moving on to the full immuno workflow.</p>
<div id="pulling-the-sequence-formatter-workflow-from-dockstore" class="section level4" number="2.4.5.1">
<h4><span class="header-section-number">2.4.5.1</span> Pulling the sequence formatter workflow from Dockstore</h4>
<p>First, visit the dockstore page for the workflow:
<!-- TODO Should version this. Should it get promoted to being official instead of in tmooney namespace? -->
* <a href="https://dockstore.org/workflows/github.com/tmooney/sequence-object-consolidator:master?tab=info">sequence-object-consolidator</a></p>
<p>At the right side of this page is a “Launch with…” section.</p>
<div class="figure">
<img src="resources/images/terra_screens/dockstore-launch_with.png" alt="" />
<p class="caption">dockstore launch with section</p>
</div>
<p>Choosing Terra opens a Terra page to select a workspace.</p>
<div class="figure">
<img src="resources/images/terra_screens/workflow-import_from_dockstore.png" alt="" />
<p class="caption">terra workflow import</p>
</div>
<p>Once the correct workspace is chosen Terra will open the Workflows tab. This is the main page for configuring the many options for a WDL workflow.</p>
</div>
<div id="selecting-the-workflow-data." class="section level4" number="2.4.5.2">
<h4><span class="header-section-number">2.4.5.2</span> Selecting the workflow data.</h4>
<p>In this case, the “Run workflow(s) with inputs defined by data table” is used to point to the data that was set up. Choose “sequence” for the table to run on.</p>
<div class="figure">
<img src="resources/images/terra_screens/workflows-select_sequence_table.png" alt="" />
<p class="caption">select table</p>
</div>
<p>Next there is a “SELECT DATA” button. In the dialogue, select all four of the sequence entities.</p>
<div class="figure">
<img src="resources/images/terra_screens/workflows-select_sequence_data.png" alt="" />
<p class="caption">select table data</p>
</div>
<p>This will automatically define a sequence set containing these four items. You can give it a custom name if desired or accept the default.</p>
</div>
<div id="setting-the-workflow-options." class="section level4" number="2.4.5.3">
<h4><span class="header-section-number">2.4.5.3</span> Setting the workflow options.</h4>
<p>The call caching doesn’t matter for this workflow, so the default is fine. There are no intermediate outputs, but it’s a good idea to be in the habit of remembering to select the box to delete them, so go ahead and mark it. The other options can remain unselected–we’re not using reference data, this workflow has no steps, so there is nothing to retry, and we shouldn’t have any empty outputs possible.</p>
<div class="figure">
<img src="resources/images/terra_screens/workflows-options.png" alt="" />
<p class="caption">workflow options</p>
</div>
<p>These options reset every time, so be sure to check that they are configured appropriately before each run. Forgetting to delete intermediate outputs can be expensive over time!</p>
</div>
<div id="setting-the-workflow-inputs." class="section level4" number="2.4.5.4">
<h4><span class="header-section-number">2.4.5.4</span> Setting the workflow inputs.</h4>
<p>The inputs to this workflow match the columns in the sequence table, so each input maps to the corresponding column by prepending it with <code>this.</code>:</p>
<table>
<thead>
<tr class="header">
<th>input</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>fastq1</td>
<td>this.fastq1</td>
</tr>
<tr class="even">
<td>fastq2</td>
<td>this.fastq2</td>
</tr>
<tr class="odd">
<td>readgroup</td>
<td>this.readgroup</td>
</tr>
</tbody>
</table>
<div class="figure">
<img src="resources/images/terra_screens/workflows-sequence_consolidator_inputs.png" alt="" />
<p class="caption">workflow inputs</p>
</div>
<p>The workflow will run once per sequence entity selected above. For each run, <code>this</code> will refer to one row in the table and pull the appropriate value.</p>
</div>
<div id="setting-the-workflow-outputs." class="section level4" number="2.4.5.5">
<h4><span class="header-section-number">2.4.5.5</span> Setting the workflow outputs.</h4>
<p>This workflow has only one output, so wire it up like the inputs.</p>
<table>
<thead>
<tr class="header">
<th>output</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>sequence_data</td>
<td>this.sequence_data</td>
</tr>
</tbody>
</table>
<div class="figure">
<img src="resources/images/terra_screens/workflows-sequence_consolidator_outputs.png" alt="" />
<p class="caption">workflow outputs</p>
</div>
<p>When the workflow runs this will automatically add a column to the sequence table named “sequence_data” with the output. As with the inputs, this happens once per row to fully populate the new column.</p>
</div>
<div id="launching-the-workflows." class="section level4" number="2.4.5.6">
<h4><span class="header-section-number">2.4.5.6</span> Launching the workflows.</h4>
<p>Since four rows were selected from the sequence table, this will launch four workflows–one per row. A dialogue after clicking “RUN ANALYSIS” will confirm this number. And then it will launch!</p>
<div class="figure">
<img src="resources/images/terra_screens/workflows-confirm_launch.png" alt="" />
<p class="caption">confirm launch</p>
</div>
</div>
<div id="seeing-the-results." class="section level4" number="2.4.5.7">
<h4><span class="header-section-number">2.4.5.7</span> Seeing the results.</h4>
<p>These workflows complete quickly so you should soon receive an e-mail with the success or failure status. Assuming they succeeded, returning to the Data tab and selecting the sequence table should show the additional column with the sequence data objects properly formatted for the <code>immuno.wdl</code> workflow.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-sequence_with_new_column.png" alt="" />
<p class="caption">sequence data with new column</p>
</div>
</div>
</div>
</div>
<div id="setting-up-the-immuno.wdl-workflow." class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Setting up the <code>immuno.wdl</code> workflow.</h2>
<p>Now that all the input data is ready, it’s time to do the configure the <code>immuno.wdl</code> workflow.</p>
<div id="pulling-the-immuno.wdl-workflow-from-dockstore" class="section level3" number="2.5.1">
<h3><span class="header-section-number">2.5.1</span> Pulling the <code>immuno.wdl</code> workflow from Dockstore</h3>
<p>First visit the dockstore page for the workflow:
<!-- TODO do we want to link to specific version or let dockstore default prevail? -->
* <a href="https://dockstore.org/workflows/github.com/wustl-oncology/analysis-wdls/immuno">immuno.wdl</a></p>
<p>As before, choose the “Launch with” option for “Terra”.</p>
</div>
<div id="selecting-the-workflow-data.-1" class="section level3" number="2.5.2">
<h3><span class="header-section-number">2.5.2</span> Selecting the workflow data.</h3>
<p>As before, choose “Run workflow(s) with inputs defined by data table”, but this time choose the “analysis” table. In the “SELECT DATA” dialogue choose the one and only row, for HCC1395.</p>
</div>
<div id="setting-the-workflow-options.-1" class="section level3" number="2.5.3">
<h3><span class="header-section-number">2.5.3</span> Setting the workflow options.</h3>
<p>Call caching can be turned on if desired–should the workflow fail this will allow it to shortcut previously successful steps. Once again it’s a good idea to select the “Delete intermediate outputs” option–though note that call caching won’t work once a run completes and the intermediate outputs have been deleted. The other options can remain unselected.</p>
</div>
<div id="setting-the-workflow-inputs.-1" class="section level3" number="2.5.4">
<h3><span class="header-section-number">2.5.4</span> Setting the workflow inputs.</h3>
<div id="static-inputs" class="section level4" number="2.5.4.1">
<h4><span class="header-section-number">2.5.4.1</span> Static inputs</h4>
<p>This time there are far more inputs to deal with. Thankfully, most of them are static values. If starting from scratch there is a link to download a JSON template that can be filled in and re-uploaded. An existing JSON from a manual run works, too. For this example, there’s <a href="https://github.com/wustl-oncology/immuno_gcp_wdl_compute1/blob/main/example_yamls/human_GRCh38_ens105/hcc1395_immuno_cloud-WDL.yaml">an existing YAML</a> that can be used. It will need to be converted to JSON for use in Terra. There are many ways to do this conversion; one is an online tool like <a href="https://jsonformatter.org/yaml-to-json">this one</a>. Using the “Upload JSON” we’ll use this to fill in most of the inputs.</p>
</div>
<div id="linking-to-the-data-table" class="section level4" number="2.5.4.2">
<h4><span class="header-section-number">2.5.4.2</span> Linking to the data table</h4>
<p>Several inputs need link to the data table, so these will need to be filled in to replace the values from the uploaded JSON (if any were set).</p>
<table>
<thead>
<tr class="header">
<th>input</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>normal_sample_name</td>
<td>this.normal_dna_sample.sample_id</td>
</tr>
<tr class="even">
<td>normal_sequence</td>
<td>this.normal_dna_sequences.sequence_data</td>
</tr>
<tr class="odd">
<td>rna_sequence</td>
<td>this.tumor_rna_sequences.sequence_data</td>
</tr>
<tr class="even">
<td>sample_name</td>
<td>this.tumor_rna_sample.sample_id</td>
</tr>
<tr class="odd">
<td>tumor_sample_name</td>
<td>this.tumor_dna_sample.sample_id</td>
</tr>
<tr class="even">
<td>tumor_sequence</td>
<td>this.tumor_dna_sequences.sequence_data</td>
</tr>
</tbody>
</table>
<p>This shows how when a table contains references to another one, you can refer to columns of the other tables through the relationships established.</p>
</div>
</div>
<div id="setting-the-workflow-outputs.-1" class="section level3" number="2.5.5">
<h3><span class="header-section-number">2.5.5</span> Setting the workflow outputs.</h3>
<p>Every output from the workflow can be added as a column to the datatable. Choose “Use defaults” to automatically populated all the outputs to columns of the same name.</p>
</div>
<div id="launching-the-workflow" class="section level3" number="2.5.6">
<h3><span class="header-section-number">2.5.6</span> Launching the workflow</h3>
<p>Now we’re ready to “Launch Analysis”. This time, we only have one row so it will launch a single workflow to do the work. For this immuno.wdl, it will take a few days to run to completion on the HCC1395, so now we wait. Terra should send an e-mail once the workflow either succeeds or fails.</p>
</div>
<div id="viewing-the-results" class="section level3" number="2.5.7">
<h3><span class="header-section-number">2.5.7</span> Viewing the results</h3>
<p>After the run, there are several ways to examine the results.</p>
<div id="in-the-data-table" class="section level4" number="2.5.7.1">
<h4><span class="header-section-number">2.5.7.1</span> In the data table</h4>
<p>The outputs that were specified will be added as columns to the “analysis” datatable.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-analysis_full.png" alt="" />
<p class="caption">filled in analysis data table</p>
</div>
</div>
<div id="from-the-command-line" class="section level4" number="2.5.7.2">
<h4><span class="header-section-number">2.5.7.2</span> From the command-line</h4>
<p>This table is now very wide! It may be useful to select the row and export it as a TSV to view.</p>
<div class="figure">
<img src="resources/images/terra_screens/data-export_to_tsv.png" alt="" />
<p class="caption">export to tsv</p>
</div>
<p>There is also a command-line tool to pull data from a terra workspace. To pull the analysis table:</p>
<pre><code>pip install firecloud
fissfc entity_tsv -p TERRA_BILLING_PROJECT -w WORKSPACE -t analysis -m flexible</code></pre>
<p>For more about this command, see <a href="https://support.terra.bio/hc/en-us/articles/360042259232-How-to-Manage-data-with-the-FISS-API">this Terra documentation page</a>.</p>
</div>
<div id="in-the-job-history" class="section level4" number="2.5.7.3">
<h4><span class="header-section-number">2.5.7.3</span> In the Job History</h4>
<p>For a more comprehensive view of what happened in the workflow, the Job History view can be used.</p>
<div class="figure">
<img src="resources/images/terra_screens/jobhistory-overview.png" alt="" />
<p class="caption">job history listings</p>
</div>
<p>This will have the attempts to run both the immuno and sequence-object-consolidator workflows. Clicking on the immuno workflow will load a details page for digging in.</p>
<div class="figure">
<img src="resources/images/terra_screens/jobhistory-details.png" alt="" />
<p class="caption">job history details</p>
</div>
<p>The “Job Manager” link on this page will show the inputs, outputs, and timings of each step in the workflow and the “Execution Directory” link will lead to the Google bucket where cromwell did its work. If intermediate outputs were not set to be deleted, they will also be in this location along with the final results.</p>
<p>For more on the job manager, see <a href="https://support.terra.bio/hc/en-us/articles/360037096272-Job-History-overview-monitoring-workflows-">this Terra documentation page</a>.</p>
</div>
</div>
</div>
<div id="further-information" class="section level2" number="2.6">
<h2><span class="header-section-number">2.6</span> Further information</h2>
<p>Terra maintains <a href="https://support.terra.bio/hc/en-us/categories/360001399872-Documentation">extensive documentation</a> including some video walkthroughs of the basics of using various features of the platform.</p>

</div>
</div>
<hr>
<center> 
  <div class="footer">
      All illustrations <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY. </a>
      <br>
      All other materials <a href= "https://creativecommons.org/licenses/by/4.0/"> CC-BY </a> unless noted otherwise.
  </div>
</center>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
