<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Immuno Workflow Course – Terra Walkthrough</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />
<link rel="shortcut icon" href="resources/images/favicon.ico" />
 <!--- go to https://favicon.io/favicon-converter/ to upload an image to make a new favicon.io. You will need to replace the current favicon.io image with the one in the downloaded directory from the website. The current image is in the resources/images/ directory --->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Immuno Workflow Course</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="terra_walkthrough.html">Terra Walkthrough</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore"><strong>Immuno Workflow Course – Terra
Walkthrough</strong></h1>

</div>

<div id="TOC">
<ul>
<li><a href="#running-a-workflow-in-terra"
id="toc-running-a-workflow-in-terra">Running a Workflow in Terra</a>
<ul>
<li><a href="#prerequisites"
id="toc-prerequisites">Prerequisites</a></li>
<li><a href="#workspace" id="toc-workspace">Workspace</a></li>
<li><a href="#adding-data" id="toc-adding-data">Adding Data</a>
<ul>
<li><a href="#uploading-via-terra-ui"
id="toc-uploading-via-terra-ui">Uploading via Terra UI</a></li>
<li><a href="#uploading-via-command-line"
id="toc-uploading-via-command-line">Uploading via Command Line</a></li>
<li><a href="#a-note-about-sharing-data-across-workspaces."
id="toc-a-note-about-sharing-data-across-workspaces.">A note about
sharing data across workspaces.</a></li>
</ul></li>
<li><a href="#populating-terra-tables"
id="toc-populating-terra-tables">Populating Terra Tables</a>
<ul>
<li><a href="#adding-the-sequence-table."
id="toc-adding-the-sequence-table.">Adding the “sequence”
table.</a></li>
<li><a href="#adding-the-sample-table."
id="toc-adding-the-sample-table.">Adding the Sample table.</a></li>
<li><a href="#adding-the-analysis-table."
id="toc-adding-the-analysis-table.">Adding the Analysis table.</a></li>
<li><a href="#linking-to-sequences-in-the-analysis-table."
id="toc-linking-to-sequences-in-the-analysis-table.">Linking to
sequences in the analysis table.</a></li>
<li><a href="#formatting-a-sequence-entry-in-the-sequence-table."
id="toc-formatting-a-sequence-entry-in-the-sequence-table.">Formatting a
sequence entry in the sequence table.</a></li>
</ul></li>
<li><a href="#setting-up-the-immuno.wdl-workflow."
id="toc-setting-up-the-immuno.wdl-workflow.">Setting up the
<code>immuno.wdl</code> workflow.</a>
<ul>
<li><a href="#pulling-the-immuno.wdl-workflow-from-dockstore"
id="toc-pulling-the-immuno.wdl-workflow-from-dockstore">Pulling the
<code>immuno.wdl</code> workflow from Dockstore</a></li>
<li><a href="#selecting-the-workflow-data.-1"
id="toc-selecting-the-workflow-data.-1">Selecting the workflow
data.</a></li>
<li><a href="#setting-the-workflow-options.-1"
id="toc-setting-the-workflow-options.-1">Setting the workflow
options.</a></li>
<li><a href="#setting-the-workflow-inputs.-1"
id="toc-setting-the-workflow-inputs.-1">Setting the workflow
inputs.</a></li>
<li><a href="#setting-the-workflow-outputs.-1"
id="toc-setting-the-workflow-outputs.-1">Setting the workflow
outputs.</a></li>
<li><a href="#launching-the-workflow"
id="toc-launching-the-workflow">Launching the workflow</a></li>
<li><a href="#viewing-the-results" id="toc-viewing-the-results">Viewing
the results</a></li>
</ul></li>
<li><a href="#further-information" id="toc-further-information">Further
information</a></li>
</ul></li>
</ul>
</div>

<div id="running-a-workflow-in-terra" class="section level1">
<h1>Running a Workflow in Terra</h1>
<div id="prerequisites" class="section level2">
<h2>Prerequisites</h2>
<p>To complete this walkthrough you’ll need to have access to a billing
account in GCP. The <a
href="https://github.com/wustl-oncology/cloud-workflows/blob/main/docs/getting_started_gcp.md">cloud-workflows
repo has more information on setting this up</a>. Then, if you haven’t
used Terra before, you’ll have to create a Terra billing project. Terra
has <a
href="https://support.terra.bio/hc/en-us/articles/360026182251">detailed
documentation on setting this up</a>.</p>
</div>
<div id="workspace" class="section level2">
<h2>Workspace</h2>
<p>All work in Terra happens in a workspace. Your workspaces are listed
on the <a href="https://app.terra.bio/#workspaces">workspaces page</a>.
It also has a button to create a workspace–the plus to the right of the
word Workspaces.</p>
<div class="float">
<img src="resources/images/terra_screens/workspaces-heading.png"
alt="Workspaces heading with + button" />
<div class="figcaption">Workspaces heading with + button</div>
</div>
<p>A unique and memorable name will make it easier to find later. This
walkthrough will use “demonstrating-immuno-workflow-in-terra”. After
choosing a name, a billing project must be selected.</p>
<div class="float">
<img src="resources/images/terra_screens/workspace-create_form.png"
alt="terra workspace form" />
<div class="figcaption">terra workspace form</div>
</div>
</div>
<div id="adding-data" class="section level2">
<h2>Adding Data</h2>
<p>The immuno workflow requires BAM or FASTQ files to be available as
inputs. This walkthrough will use the HCC1395 data available here:</p>
<ul>
<li><a
href="http://genomedata.org/pmbio-workshop/fastqs/all/Exome_Norm.tar">HCC1395
Normal</a></li>
<li><a
href="http://genomedata.org/pmbio-workshop/fastqs/all/Exome_Tumor.tar">HCC1395
Tumor</a></li>
<li><a
href="http://genomedata.org/pmbio-workshop/fastqs/all/RNAseq_Tumor.tar">HCC1395
Tumor RNA</a></li>
</ul>
<p>These files will need to be extracted and uploaded to Terra. First
extract all of these tar files to a <code>sequence_data</code>
directory. Then, there are a few ways to upload that directory to
Terra:</p>
<div id="uploading-via-terra-ui" class="section level3">
<h3>Uploading via Terra UI</h3>
<p>On the “Data” tab for the workspace will be a “Files” link under
“Other Data”. From that section there is an “Upload” button that can be
used to upload the desired data.</p>
<div class="float">
<img src="resources/images/terra_screens/data-tab_files.png"
alt="terra upload form" />
<div class="figcaption">terra upload form</div>
</div>
</div>
<div id="uploading-via-command-line" class="section level3">
<h3>Uploading via Command Line</h3>
<p>On the “Dashboard” tab for the workspace will be a link to the Google
Cloud Storage Bucket for the workspace. By copying this bucket URL, the
files can be uploaded using the <code>gsutil</code> tool.</p>
<div class="float">
<img
src="resources/images/terra_screens/dashboard-cloud_information.png"
alt="dashboard cloud information" />
<div class="figcaption">dashboard cloud information</div>
</div>
</div>
<div id="a-note-about-sharing-data-across-workspaces."
class="section level3">
<h3>A note about sharing data across workspaces.</h3>
<p>If the same data will be used by many people, it may be beneficial to
upload it to a workspace and configure it once. Thereafter that
workspace can be cloned to use the data in another workspace without
copying it again thereby saving storage costs. The data tables
configured below could also be cloned to the new workspace.</p>
</div>
</div>
<div id="populating-terra-tables" class="section level2">
<h2>Populating Terra Tables</h2>
<p>For this example, there will be three data tables: “sequence” (to
list the information for each pair of input FASTQs), “sample” (for any
metadata about the samples), and “analysis” (to group together the
sequences and samples for one workflow run). In the simplest case a
workflow can be run from a single table (or by supplying an input JSON
directly); the additional tables here are to demonstrate some of the
linking features of Terra.</p>
<div id="adding-the-sequence-table." class="section level3">
<h3>Adding the “sequence” table.</h3>
<p>On the “Data” tab of the workspace is an “Import Data” button.
Clicking on it offers a few options–in this case use the import TSV
option.</p>
<div class="float">
<img src="resources/images/terra_screens/data-upload_tsv.png"
alt="import menu" />
<div class="figcaption">import menu</div>
</div>
<p>This offers the option to upload a file or to copy and paste the TSV.
To create the “sequence” table, start with the following:</p>
<pre><code>entity:sequence_id  sample_id   fastq1  fastq2  readgroup
1   HCC1395_TUMOR_DNA   gs://BUCKET/sequence_data/Exome_Tumor/Exome_Tumor_R1.fastq.gz   gs://BUCKET/sequence_data/Exome_Tumor/Exome_Tumor_R2.fastq.gz   @RG\tID:1\tPU:C1TD1ACXX\tSM:HCC1395_TUMOR_DNA\tLB:HCC1395_TUMOR_DNA_LIB1\tPL:Illumina\tCN:WUGSC
2   HCC1395_NORMAL_DNA  gs://BUCKET/sequence_data/Exome_Norm/Exome_Norm_R1.fastq.gz gs://BUCKET/sequence_data/Exome_Norm/Exome_Norm_R2.fastq.gz @RG\tID:2\tPU:C1TD1ACXX\tSM:HCC1395_NORMAL_DNA\tLB:HCC1395_NORMAL_DNA_LIB1\tPL:Illumina\tCN:WUGSC
3   HCC1395_TUMOR_RNA   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane1_R1.fastq.gz   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane1_R2.fastq.gz   &quot;ID:3   PU:H3MYFBBXX.4  SM:HCC1395_TUMOR_RNA    LB:HCC1395_TUMOR_RNA_LIB1   PL:Illumina CN:WUGSC&quot;
4   HCC1395_TUMOR_RNA   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane2_R1.fastq.gz   gs://BUCKET/sequence_data/RNAseq_Tumor/RNAseq_Tumor_Lane2_R2.fastq.gz   &quot;ID:4   PU:H3MYFBBXX.5  SM:HCC1395_TUMOR_RNA    LB:HCC1395_TUMOR_RNA_LIB1   PL:Illumina CN:WUGSC&quot;</code></pre>
<p>The bucket paths need to be updated to point to the uploaded files
for this workspace by filling in the <code>BUCKET</code>. Notice the
headers of this file… the first column starts with <code>entity:</code>
to indicate this column identifies the table. Since it says
<code>sequence_id</code> the table will be named “sequence”. For this
table, the file import must be used as the text uploader does not handle
fields that contain tabs.</p>
<div class="float">
<img src="resources/images/terra_screens/data-import_file.png"
alt="import file" />
<div class="figcaption">import file</div>
</div>
</div>
<div id="adding-the-sample-table." class="section level3">
<h3>Adding the Sample table.</h3>
<p>This time the text import can be used by copying this block:</p>
<pre><code>entity:sample_id    individual_id   type
HCC1395_TUMOR_DNA   HCC1395 tumor_dna
HCC1395_NORMAL_DNA  HCC1395 normal_dna
HCC1395_TUMOR_RNA   HCC1395 tumor_rna</code></pre>
<p>This will create a “sample” table.</p>
<div class="float">
<img src="resources/images/terra_screens/data-import_text.png"
alt="import text" />
<div class="figcaption">import text</div>
</div>
</div>
<div id="adding-the-analysis-table." class="section level3">
<h3>Adding the Analysis table.</h3>
<p>As before, open the TSV import and copy this block:</p>
<pre><code>entity:analysis_id  individual_id   tumor_dna_sample    normal_dna_sample   tumor_rna_sample
HCC1395 HCC1395 {&quot;entityType&quot;:&quot;sample&quot;,&quot;entityName&quot;:&quot;HCC1395_NORMAL_DNA&quot;}   {&quot;entityType&quot;:&quot;sample&quot;,&quot;entityName&quot;:&quot;HCC1395_TUMOR_DNA&quot;}    {&quot;entityType&quot;:&quot;sample&quot;,&quot;entityName&quot;:&quot;HCC1395_TUMOR_RNA&quot;}</code></pre>
<p>Notice this table has some entities in the columns. This will
automatically link those entries to the “sample” table. This allows
access to properties of the “sample” table via this table. If there were
another table for individuals, the “sample” table could’ve been created
with links to that table for further nested indirect access.</p>
</div>
<div id="linking-to-sequences-in-the-analysis-table."
class="section level3">
<h3>Linking to sequences in the analysis table.</h3>
<p>Now that the basic tables are in place, there are a few more steps to
do before using the workflow. First is to add columns to the analysis
table to point to which sequence applies. From the “analysis” table,
there is an edit button with an option to “Add Column”. Use it three
times, one for each sample.</p>
<div class="float">
<img src="resources/images/terra_screens/data-edit_analysis_table.png"
alt="edit analysis table" />
<div class="figcaption">edit analysis table</div>
</div>
<p>Each time, use a type of “reference” to the “sequence” table and
select the option that the values are a list.</p>
<p>The three columns to add and their values:</p>
<table>
<thead>
<tr class="header">
<th>column</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>tumor_dna_sequences</td>
<td>1</td>
</tr>
<tr class="even">
<td>normal_dna_sequences</td>
<td>2</td>
</tr>
<tr class="odd">
<td>tumor_rna_sequences</td>
<td>3,4</td>
</tr>
</tbody>
</table>
<p>The option to use a list is selected every time because the workflow
expects a list of data, even if some lists only contain one item.</p>
<div class="float">
<img src="resources/images/terra_screens/data-add_analysis_column.png"
alt="add column form" />
<div class="figcaption">add column form</div>
</div>
<p>As a note, although this walkthrough used the Terra UI to demonstrate
adding a column to the workflow, it would’ve been possible to create
these entities in the TSV directly and skip this step so long as the
column is formatted correctly. An example of how to format a list in the
TSV:</p>
<pre><code>[{&quot;entityType&quot;:&quot;sequence&quot;,&quot;entityName&quot;:&quot;1&quot;},{&quot;entityType&quot;:&quot;sequence&quot;,&quot;entityName&quot;:&quot;2&quot;}]</code></pre>
</div>
<div id="formatting-a-sequence-entry-in-the-sequence-table."
class="section level3">
<h3>Formatting a sequence entry in the sequence table.</h3>
<p>The Terra interface does not support adding a JSON object directly as
a single column unless that object is the output of a workflow.
Therefore a separate zero-step WDL workflow is needed in order to format
the sequence objects in the way the <code>immuno.wdl</code> expects.
This is a great opportunity to see a simple case of setting up a
workflow before moving on to the full immuno workflow.</p>
<div id="pulling-the-sequence-formatter-workflow-from-dockstore"
class="section level4">
<h4>Pulling the sequence formatter workflow from Dockstore</h4>
<p>First, visit the dockstore page for the workflow:
<!-- TODO Should version this. Should it get promoted to being official instead of in tmooney namespace? -->
* <a
href="https://dockstore.org/workflows/github.com/tmooney/sequence-object-consolidator:master?tab=info">sequence-object-consolidator</a></p>
<p>At the right side of this page is a “Launch with…” section.</p>
<div class="float">
<img src="resources/images/terra_screens/dockstore-launch_with.png"
alt="dockstore launch with section" />
<div class="figcaption">dockstore launch with section</div>
</div>
<p>Choosing Terra opens a Terra page to select a workspace.</p>
<div class="float">
<img
src="resources/images/terra_screens/workflow-import_from_dockstore.png"
alt="terra workflow import" />
<div class="figcaption">terra workflow import</div>
</div>
<p>Once the correct workspace is chosen Terra will open the Workflows
tab. This is the main page for configuring the many options for a WDL
workflow.</p>
</div>
<div id="selecting-the-workflow-data." class="section level4">
<h4>Selecting the workflow data.</h4>
<p>In this case, the “Run workflow(s) with inputs defined by data table”
is used to point to the data that was set up. Choose “sequence” for the
table to run on.</p>
<div class="float">
<img
src="resources/images/terra_screens/workflows-select_sequence_table.png"
alt="select table" />
<div class="figcaption">select table</div>
</div>
<p>Next there is a “SELECT DATA” button. In the dialogue, select all
four of the sequence entities.</p>
<div class="float">
<img
src="resources/images/terra_screens/workflows-select_sequence_data.png"
alt="select table data" />
<div class="figcaption">select table data</div>
</div>
<p>This will automatically define a sequence set containing these four
items. You can give it a custom name if desired or accept the
default.</p>
</div>
<div id="setting-the-workflow-options." class="section level4">
<h4>Setting the workflow options.</h4>
<p>The call caching doesn’t matter for this workflow, so the default is
fine. There are no intermediate outputs, but it’s a good idea to be in
the habit of remembering to select the box to delete them, so go ahead
and mark it. The other options can remain unselected–we’re not using
reference data, this workflow has no steps, so there is nothing to
retry, and we shouldn’t have any empty outputs possible.</p>
<div class="float">
<img src="resources/images/terra_screens/workflows-options.png"
alt="workflow options" />
<div class="figcaption">workflow options</div>
</div>
<p>These options reset every time, so be sure to check that they are
configured appropriately before each run. Forgetting to delete
intermediate outputs can be expensive over time!</p>
</div>
<div id="setting-the-workflow-inputs." class="section level4">
<h4>Setting the workflow inputs.</h4>
<p>The inputs to this workflow match the columns in the sequence table,
so each input maps to the corresponding column by prepending it with
<code>this.</code>:</p>
<table>
<thead>
<tr class="header">
<th>input</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>fastq1</td>
<td>this.fastq1</td>
</tr>
<tr class="even">
<td>fastq2</td>
<td>this.fastq2</td>
</tr>
<tr class="odd">
<td>readgroup</td>
<td>this.readgroup</td>
</tr>
</tbody>
</table>
<div class="float">
<img
src="resources/images/terra_screens/workflows-sequence_consolidator_inputs.png"
alt="workflow inputs" />
<div class="figcaption">workflow inputs</div>
</div>
<p>The workflow will run once per sequence entity selected above. For
each run, <code>this</code> will refer to one row in the table and pull
the appropriate value.</p>
</div>
<div id="setting-the-workflow-outputs." class="section level4">
<h4>Setting the workflow outputs.</h4>
<p>This workflow has only one output, so wire it up like the inputs.</p>
<table>
<thead>
<tr class="header">
<th>output</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>sequence_data</td>
<td>this.sequence_data</td>
</tr>
</tbody>
</table>
<div class="float">
<img
src="resources/images/terra_screens/workflows-sequence_consolidator_outputs.png"
alt="workflow outputs" />
<div class="figcaption">workflow outputs</div>
</div>
<p>When the workflow runs this will automatically add a column to the
sequence table named “sequence_data” with the output. As with the
inputs, this happens once per row to fully populate the new column.</p>
</div>
<div id="launching-the-workflows." class="section level4">
<h4>Launching the workflows.</h4>
<p>Since four rows were selected from the sequence table, this will
launch four workflows–one per row. A dialogue after clicking “RUN
ANALYSIS” will confirm this number. And then it will launch!</p>
<div class="float">
<img src="resources/images/terra_screens/workflows-confirm_launch.png"
alt="confirm launch" />
<div class="figcaption">confirm launch</div>
</div>
</div>
<div id="seeing-the-results." class="section level4">
<h4>Seeing the results.</h4>
<p>These workflows complete quickly so you should soon receive an e-mail
with the success or failure status. Assuming they succeeded, returning
to the Data tab and selecting the sequence table should show the
additional column with the sequence data objects properly formatted for
the <code>immuno.wdl</code> workflow.</p>
<div class="float">
<img
src="resources/images/terra_screens/data-sequence_with_new_column.png"
alt="sequence data with new column" />
<div class="figcaption">sequence data with new column</div>
</div>
</div>
</div>
</div>
<div id="setting-up-the-immuno.wdl-workflow." class="section level2">
<h2>Setting up the <code>immuno.wdl</code> workflow.</h2>
<p>Now that all the input data is ready, it’s time to do the configure
the <code>immuno.wdl</code> workflow.</p>
<div id="pulling-the-immuno.wdl-workflow-from-dockstore"
class="section level3">
<h3>Pulling the <code>immuno.wdl</code> workflow from Dockstore</h3>
<p>First visit the dockstore page for the workflow:
<!-- TODO do we want to link to specific version or let dockstore default prevail? -->
* <a
href="https://dockstore.org/workflows/github.com/wustl-oncology/analysis-wdls/immuno">immuno.wdl</a></p>
<p>As before, choose the “Launch with” option for “Terra”.</p>
</div>
<div id="selecting-the-workflow-data.-1" class="section level3">
<h3>Selecting the workflow data.</h3>
<p>As before, choose “Run workflow(s) with inputs defined by data
table”, but this time choose the “analysis” table. In the “SELECT DATA”
dialogue choose the one and only row, for HCC1395.</p>
</div>
<div id="setting-the-workflow-options.-1" class="section level3">
<h3>Setting the workflow options.</h3>
<p>Call caching can be turned on if desired–should the workflow fail
this will allow it to shortcut previously successful steps. Once again
it’s a good idea to select the “Delete intermediate outputs”
option–though note that call caching won’t work once a run completes and
the intermediate outputs have been deleted. The other options can remain
unselected.</p>
</div>
<div id="setting-the-workflow-inputs.-1" class="section level3">
<h3>Setting the workflow inputs.</h3>
<div id="static-inputs" class="section level4">
<h4>Static inputs</h4>
<p>This time there are far more inputs to deal with. Thankfully, most of
them are static values. If starting from scratch there is a link to
download a JSON template that can be filled in and re-uploaded. An
existing JSON from a manual run works, too. For this example, there’s <a
href="https://github.com/wustl-oncology/immuno_gcp_wdl_compute1/blob/main/example_yamls/human_GRCh38_ens105/hcc1395_immuno_cloud-WDL.yaml">an
existing YAML</a> that can be used. It will need to be converted to JSON
for use in Terra. There are many ways to do this conversion; one is an
online tool like <a href="https://jsonformatter.org/yaml-to-json">this
one</a>. Using the “Upload JSON” we’ll use this to fill in most of the
inputs.</p>
</div>
<div id="linking-to-the-data-table" class="section level4">
<h4>Linking to the data table</h4>
<p>Several inputs need link to the data table, so these will need to be
filled in to replace the values from the uploaded JSON (if any were
set).</p>
<table>
<thead>
<tr class="header">
<th>input</th>
<th>value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>normal_sample_name</td>
<td>this.normal_dna_sample.sample_id</td>
</tr>
<tr class="even">
<td>normal_sequence</td>
<td>this.normal_dna_sequences.sequence_data</td>
</tr>
<tr class="odd">
<td>rna_sequence</td>
<td>this.tumor_rna_sequences.sequence_data</td>
</tr>
<tr class="even">
<td>sample_name</td>
<td>this.tumor_rna_sample.sample_id</td>
</tr>
<tr class="odd">
<td>tumor_sample_name</td>
<td>this.tumor_dna_sample.sample_id</td>
</tr>
<tr class="even">
<td>tumor_sequence</td>
<td>this.tumor_dna_sequences.sequence_data</td>
</tr>
</tbody>
</table>
<p>This shows how when a table contains references to another one, you
can refer to columns of the other tables through the relationships
established.</p>
</div>
</div>
<div id="setting-the-workflow-outputs.-1" class="section level3">
<h3>Setting the workflow outputs.</h3>
<p>Every output from the workflow can be added as a column to the
datatable. Choose “Use defaults” to automatically populated all the
outputs to columns of the same name.</p>
</div>
<div id="launching-the-workflow" class="section level3">
<h3>Launching the workflow</h3>
<p>Now we’re ready to “Launch Analysis”. This time, we only have one row
so it will launch a single workflow to do the work. For this immuno.wdl,
it will take a few days to run to completion on the HCC1395, so now we
wait. Terra should send an e-mail once the workflow either succeeds or
fails.</p>
</div>
<div id="viewing-the-results" class="section level3">
<h3>Viewing the results</h3>
<p>After the run, there are several ways to examine the results.</p>
<div id="in-the-data-table" class="section level4">
<h4>In the data table</h4>
<p>The outputs that were specified will be added as columns to the
“analysis” datatable.</p>
<div class="float">
<img src="resources/images/terra_screens/data-analysis_full.png"
alt="filled in analysis data table" />
<div class="figcaption">filled in analysis data table</div>
</div>
</div>
<div id="from-the-command-line" class="section level4">
<h4>From the command-line</h4>
<p>This table is now very wide! It may be useful to select the row and
export it as a TSV to view.</p>
<div class="float">
<img src="resources/images/terra_screens/data-export_to_tsv.png"
alt="export to tsv" />
<div class="figcaption">export to tsv</div>
</div>
<p>There is also a command-line tool to pull data from a terra
workspace. To pull the analysis table:</p>
<pre><code>pip install firecloud
fissfc entity_tsv -p TERRA_BILLING_PROJECT -w WORKSPACE -t analysis -m flexible</code></pre>
<p>For more about this command, see <a
href="https://support.terra.bio/hc/en-us/articles/360042259232-How-to-Manage-data-with-the-FISS-API">this
Terra documentation page</a>.</p>
</div>
<div id="in-the-job-history" class="section level4">
<h4>In the Job History</h4>
<p>For a more comprehensive view of what happened in the workflow, the
Job History view can be used.</p>
<div class="float">
<img src="resources/images/terra_screens/jobhistory-overview.png"
alt="job history listings" />
<div class="figcaption">job history listings</div>
</div>
<p>This will have the attempts to run both the immuno and
sequence-object-consolidator workflows. Clicking on the immuno workflow
will load a details page for digging in.</p>
<div class="float">
<img src="resources/images/terra_screens/jobhistory-details.png"
alt="job history details" />
<div class="figcaption">job history details</div>
</div>
<p>The “Job Manager” link on this page will show the inputs, outputs,
and timings of each step in the workflow and the “Execution Directory”
link will lead to the Google bucket where cromwell did its work. If
intermediate outputs were not set to be deleted, they will also be in
this location along with the final results.</p>
<p>For more on the job manager, see <a
href="https://support.terra.bio/hc/en-us/articles/360037096272-Job-History-overview-monitoring-workflows-">this
Terra documentation page</a>.</p>
</div>
</div>
</div>
<div id="further-information" class="section level2">
<h2>Further information</h2>
<p>Terra maintains <a
href="https://support.terra.bio/hc/en-us/categories/360001399872-Documentation">extensive
documentation</a> including some video walkthroughs of the basics of
using various features of the platform.</p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
